src/providers-registry/providers-registry-contract.ts

```diff
@@ -0,0 +1,270 @@
+
+    // note: this is safe, as adminsPool array is initialized when provider is added
+    providers[data.providerId].adminsPool.push(...data.admins);
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  // admin-only: switch trace state
+  if (input.function === "switch-trace") {
+    checkIsContractAdmin(caller);
+    state.trace = !state.trace;
+
+    trace("END STATE", state);
+
+    return {state}
+  }
+
+  if (input.function === "add-contract-admins") {
+    const data = input.data as AddContractAdmins;
+    if (data.admins === undefined || data.admins.length === 0) {
+      throw new ContractError("New contract admins not defined");
+    }
+    checkIsContractAdmin(caller);
+    state.contractAdmins.push(...data.admins);
+
+    return {state }
+  }
+
+
+  /* STATE READ_ONLY ACTIONS */
+  // admin-only: returns last uploaded providers-registry - ie. last element in the manifests array
+  if (input.function === "provider-data") {
+    const data = input.data as GetProviderData;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+
+    return {
+      result: {
+        provider: providers[data.providerId]
+      }
+    }
+  }
+
+  if (input.function === "active-manifest") {
+    const data = input.data as GetProviderManifest;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+
+    const manifests = getManifestsForCaller(caller);
+    trace("Block height: ", blockHeight);
+    const manifest = manifests.slice().reverse().find((manifest) => {
+      trace("Checking manifest: ", manifest);
+      trace("Checking lockedBlocks: ", manifest.lockedHours * BLOCKS_IN_HOUR);
+      return manifest.uploadBlockHeight + (manifest.lockedHours * BLOCKS_IN_HOUR) <= blockHeight;
+    }) || manifests[0];
+
+    trace("RESULT", manifest);
+
+    return {result: {manifest}}
+  }
+
+  function checkIsContractAdmin(idWallet: string) {
+    if (!isContractAdmin(idWallet)) {
+      throw new ContractError("Only admin is allowed to call this function");
+    }
+  }
+
+  function isContractAdmin(idWallet: string) {
+    const wallet = verifyWalletAddress(idWallet);
+    return state.contractAdmins.includes(wallet);
+  }
+
+  function verifyWalletAddress(addr: string): string {
+    const address = addr.toString().trim();
+    if (!/[a-z0-9_-]{43}/i.test(address)) {
+      throw new ContractError('Invalid Arweave address.');
+    }
+
+    return address;
+  }
+
+  function trace(log: string, data: any = {}) {
+    if (state.trace) {
+      console.log(`\n=== ${log} ===\n`, JSON.stringify(data));
+    }
+  }
+
+  function getManifestsForCaller(providerId: string): ManifestData[] {
+    const manifests = providers[providerId].manifests;
+    if (manifests === undefined || manifests.length === 0) {
+      throw new ContractError(`No manifests defined for ${caller}`);
+    }
+    return manifests;
+  }
+
+  function checkPrivileges(caller: string, providerId: string) {
```

Comment:
mustBeProvidersAdmin ?

Reply:
as I wrote, I'm used to a naming convention, where "check(...)" functions throw exceptions, when the conditions is false. So if I see anything that starts with "check" I know it can throw an exception (ie. runtime exception in Java world).